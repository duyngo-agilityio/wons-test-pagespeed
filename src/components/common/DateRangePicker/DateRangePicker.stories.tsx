import type { Meta } from '@storybook/react';

// Components
import DateRangePicker from './index';

const meta: Meta<typeof DateRangePicker> = {
  title: 'Components/Common/DateRangePicker',
  component: DateRangePicker,
  parameters: {
    layout: 'centered',
    controls: { expanded: true },
  },
  tags: ['autodocs'],
  argTypes: {
    title: {
      description: 'Text content',
    },
    size: {
      description: 'Size of text',
    },
    as: {
      description: 'HTML tag',
    },
    className: {
      description: 'Additional class name',
    },
  },
} satisfies Meta<typeof DateRangePicker>;

export default meta;

export const Default = () => (
  <div className="w-[320px] p-[20px] bg-gray-50 dark:bg-gray-600 rounded-2xl">
    <DateRangePicker />
  </div>
);
