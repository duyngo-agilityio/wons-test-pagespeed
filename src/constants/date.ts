export const DEFAULT_RANGE_VALUE_PICKER = 7;
