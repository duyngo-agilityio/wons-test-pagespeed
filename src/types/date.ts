import { DateValue } from '@internationalized/date';

export interface DateRangeState {
  start: DateValue;
  end: DateValue;
}
