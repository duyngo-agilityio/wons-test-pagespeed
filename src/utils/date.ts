import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import {
  CalendarDate,
  getLocalTimeZone,
  parseDate,
  today,
} from '@internationalized/date';

// Types
import { DateRangeState } from '@/types';

// Constants
import { DAYJS_PATTERN } from '@/constants';

dayjs.extend(utc);

/***
 * Function format date
 */
export const formatDate = (date: string | Date, pattern: string) =>
  dayjs(date).utc(true).format(pattern);

/**
 * Function get start time for DatePicker
 */
export const getStartTimeDatePicker = (
  date: number,
  startTime: string,
  endTime: string,
): DateRangeState => {
  // Get the current date in the local timezone
  const currentDate = today(getLocalTimeZone());
  const defaultStartTime = currentDate.add({
    days: -date,
  });
  const defaultStartTimeIso = `${defaultStartTime.year}-${String(
    defaultStartTime.month,
  ).padStart(2, '0')}-${String(defaultStartTime.day).padStart(2, '0')}`;

  return {
    start: parseDate(
      startTime
        ? formatDate(new Date(startTime), DAYJS_PATTERN['YYYY-MM-DD'])
        : defaultStartTimeIso,
    ),

    end: endTime
      ? parseDate(formatDate(new Date(endTime), DAYJS_PATTERN['YYYY-MM-DD']))
      : currentDate,
  };
};

/**
 * Function Get current date
 */
export const currentDate = today(getLocalTimeZone());

export const formatDateByISO = (date: string): string =>
  dayjs(date).utc(true).format();

export const convertToCalendarDate = (
  date?: string,
): CalendarDate | undefined => {
  if (!date) {
    return undefined;
  }

  const [year, month, day] = date.split('-');

  return new CalendarDate(Number(year), Number(month), Number(day));
};
