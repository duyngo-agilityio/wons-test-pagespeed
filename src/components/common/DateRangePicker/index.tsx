'use client';

import React, { useState, useCallback } from 'react';
import { getLocalTimeZone } from '@internationalized/date';
import {
  DateRangePicker as DateRangePickerBase,
  DateRangePickerProps,
  DateValue,
  RangeValue,
} from '@nextui-org/react';

// Utils
import { currentDate, formatDate, getStartTimeDatePicker } from '@/utils';

// Types
import { DateRangeState } from '@/types';

// Icons
import { FaChevronDown } from 'react-icons/fa6';

interface DateRangePickerBaseProps extends DateRangePickerProps {}

const DateRangePicker = ({
  minValue,
  maxValue,
  ...rest
}: DateRangePickerBaseProps) => {
  const [dateRange, setDateRange] = useState<DateRangeState | null>(
    getStartTimeDatePicker(6),
  );

  const [isOpenRangePicker, setIsOpenRangePicker] = useState(false);

  const handleOpenDatePicker = useCallback(
    () => setIsOpenRangePicker((prev) => !prev),
    [setIsOpenRangePicker],
  );

  const handleOnChangeDateRangePicker = useCallback(
    (value: RangeValue<DateValue>) => {
      setIsOpenRangePicker(false);
      setDateRange(value);
    },
    [setDateRange, setIsOpenRangePicker],
  );

  return (
    <div className="max-w-[290.5px] relative">
      <div className="flex gap-[15.85px]">
        {dateRange &&
          Object.keys(dateRange).map((key) => (
            <div
              key={key}
              className="flex w-full relative bg-white dark:bg-gray-400 : rounded-[5.28px] h-[42.25px] items-center"
            >
              <p className="text-[15.02px] ml-[17.96px]">
                {formatDate(
                  dateRange?.[key as keyof typeof dateRange].toDate(
                    getLocalTimeZone(),
                  ),
                )}
              </p>
              <div className="absolute flex items-center h-full right-0 px-2">
                <button
                  className="flex h-[15px] w-[15px] justify-center items-center rounded-full hover:bg-gray-100"
                  onClick={handleOpenDatePicker}
                >
                  <FaChevronDown className="w-[8.43px]" />
                </button>
              </div>
            </div>
          ))}
      </div>
      <div className="absolute z-[-1] w-full top-0">
        <DateRangePickerBase
          classNames={{
            innerWrapper: 'z-[-1]',
          }}
          className="h-[42.25px]"
          isOpen={isOpenRangePicker}
          value={dateRange}
          label="Date Range Picker"
          minValue={minValue}
          maxValue={maxValue ?? currentDate}
          onChange={handleOnChangeDateRangePicker}
          {...rest}
        />
      </div>
    </div>
  );
};

export default DateRangePicker;
