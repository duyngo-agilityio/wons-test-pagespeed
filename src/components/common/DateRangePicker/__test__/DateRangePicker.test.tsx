import { DateRangePickerProps } from '@nextui-org/react';
import { parseAbsoluteToLocal } from '@internationalized/date';

// Components
import DateRangePicker from '../index';

const mockOnClick = jest.fn();

describe('DateRangePicker', () => {
  const renderComponent = (props?: DateRangePickerProps) =>
    testLibJestUtils.render(
      <DateRangePicker
        onClick={mockOnClick}
        maxValue={parseAbsoluteToLocal('2024-09-07T18:45:22Z')}
        {...props}
      />,
    );

  afterEach(() => {
    mockOnClick.mockClear();
  });

  it('match snapshot', () => {
    const { container } = renderComponent();

    expect(container).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
