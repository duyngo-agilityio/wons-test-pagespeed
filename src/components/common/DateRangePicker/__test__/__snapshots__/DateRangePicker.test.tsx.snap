// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateRangePicker match snapshot 1`] = `
<div>
  <div
    data-overlay-container="true"
  >
    <script>
      !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&false)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}else{c.add('light')}if(e==='light'||e==='dark'||!e)d.style.colorScheme=e||'light'}catch(e){}}()
    </script>
    <div
      class="w-full md:max-w-[290.5px] relative bg-gray-50 dark:bg-gray-600 mt-8 md:mt-0"
    >
      <div
        class="flex justify-between md:gap-[15.85px]"
      >
        <div
          class="flex max-w-36 w-36 relative bg-white cursor-pointer dark:bg-gray-400 rounded-[5.28px] h-[42.25px] items-center"
        >
          <p
            class="text-[15.02px] ml-[17.96px]"
          >
            24-10-2024
          </p>
          <div
            class="absolute flex items-center h-full right-0 px-2"
          >
            <button
              aria-label="start date"
              class="flex h-[15px] w-[15px] justify-center items-center rounded-full hover:bg-gray-100"
            >
              <svg
                class="w-[8.43px]"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="flex max-w-36 w-36 relative bg-white cursor-pointer dark:bg-gray-400 rounded-[5.28px] h-[42.25px] items-center"
        >
          <p
            class="text-[15.02px] ml-[17.96px]"
          >
            30-10-2024
          </p>
          <div
            class="absolute flex items-center h-full right-0 px-2"
          >
            <button
              aria-label="end date"
              class="flex h-[15px] w-[15px] justify-center items-center rounded-full hover:bg-gray-100"
            >
              <svg
                class="w-[8.43px]"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="absolute z-[-1] w-full top-0"
      >
        <div
          class="group flex flex-col w-full h-[42.25px]"
          data-has-helper="true"
          data-invalid="true"
          data-slot="base"
        >
          <div
            aria-describedby="react-aria-description-2 react-aria-:r4:"
            aria-labelledby="react-aria-:r1:"
            class="relative px-3 inline-flex cursor-text tap-highlight-transparent bg-default-100 hover:bg-default-200 focus-within:hover:bg-default-100 group-data-[invalid=true]:bg-danger-50 group-data-[invalid=true]:hover:bg-danger-100 group-data-[invalid=true]:focus-within:hover:bg-danger-50 min-h-10 rounded-medium flex-col items-start justify-center gap-0 w-full h-14 py-2 shadow-none"
            data-slot="input-wrapper"
            id="react-aria-:r0:"
            role="group"
          >
            <span
              class="block subpixel-antialiased text-default-600 group-data-[required=true]:after:content-['*'] group-data-[required=true]:after:text-danger group-data-[required=true]:after:ml-0.5 group-data-[invalid=true]:text-danger w-full text-tiny cursor-text"
              data-slot="label"
              id="react-aria-:r1:"
            >
              Date Range Picker
            </span>
            <div
              class="flex items-center text-default-400 w-full gap-x-2 h-6 group-data-[invalid=true]:text-danger z-[-1]"
              data-slot="inner-wrapper"
            >
              <div
                aria-describedby="react-aria-description-2 react-aria-:r4:"
                aria-label="Start Date"
                aria-labelledby="react-aria-:r1:"
                class="flex h-full gap-x-0.5 w-full font-normal text-small"
                data-slot="start-input"
                role="presentation"
                style="max-width: fit-content;"
                value="2024-10-24"
              >
                <div
                  aria-describedby="react-aria-description-2 react-aria-:r4:"
                  aria-invalid="true"
                  aria-label="month, Start Date, "
                  aria-labelledby="react-aria-:rc: react-aria-:r1:"
                  aria-valuemax="12"
                  aria-valuemin="1"
                  aria-valuenow="10"
                  aria-valuetext="10 â€“ October"
                  autocorrect="off"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  contenteditable="true"
                  data-editable="true"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="month"
                  enterkeyhint="next"
                  id="react-aria-:rc:"
                  inputmode="numeric"
                  role="spinbutton"
                  spellcheck="false"
                  style="caret-color: transparent;"
                  tabindex="0"
                >
                  10
                </div>
                <div
                  aria-hidden="true"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="literal"
                >
                  /
                </div>
                <div
                  aria-describedby="react-aria-description-2 react-aria-:r4:"
                  aria-invalid="true"
                  aria-label="day, Start Date, "
                  aria-labelledby="react-aria-:rg: react-aria-:r1:"
                  aria-valuemax="31"
                  aria-valuemin="1"
                  aria-valuenow="24"
                  aria-valuetext="24"
                  autocorrect="off"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  contenteditable="true"
                  data-editable="true"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="day"
                  enterkeyhint="next"
                  id="react-aria-:rg:"
                  inputmode="numeric"
                  role="spinbutton"
                  spellcheck="false"
                  style="caret-color: transparent;"
                  tabindex="0"
                >
                  24
                </div>
                <div
                  aria-hidden="true"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="literal"
                >
                  /
                </div>
                <div
                  aria-describedby="react-aria-description-2 react-aria-:r4:"
                  aria-invalid="true"
                  aria-label="year, Start Date, "
                  aria-labelledby="react-aria-:rk: react-aria-:r1:"
                  aria-valuemax="9999"
                  aria-valuemin="1"
                  aria-valuenow="2024"
                  aria-valuetext="2024"
                  autocorrect="off"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  contenteditable="true"
                  data-editable="true"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="year"
                  enterkeyhint="next"
                  id="react-aria-:rk:"
                  inputmode="numeric"
                  role="spinbutton"
                  spellcheck="false"
                  style="caret-color: transparent;"
                  tabindex="0"
                >
                  2024
                </div>
                <input
                  type="hidden"
                  value="2024-10-24"
                />
              </div>
              <span
                aria-hidden="true"
                class="-mx-1 text-inherit"
                data-slot="separator"
                role="separator"
              >
                -
              </span>
              <div
                aria-describedby="react-aria-description-2 react-aria-:r4:"
                aria-label="End Date"
                aria-labelledby="react-aria-:r1:"
                class="flex h-full gap-x-0.5 w-full font-normal text-small"
                data-slot="end-input"
                role="presentation"
                value="2024-10-30"
              >
                <div
                  aria-describedby="react-aria-description-2 react-aria-:r4:"
                  aria-invalid="true"
                  aria-label="month, End Date, "
                  aria-labelledby="react-aria-:rr: react-aria-:r1:"
                  aria-valuemax="12"
                  aria-valuemin="1"
                  aria-valuenow="10"
                  aria-valuetext="10 â€“ October"
                  autocorrect="off"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  contenteditable="true"
                  data-editable="true"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="month"
                  enterkeyhint="next"
                  id="react-aria-:rr:"
                  inputmode="numeric"
                  role="spinbutton"
                  spellcheck="false"
                  style="caret-color: transparent;"
                  tabindex="0"
                >
                  10
                </div>
                <div
                  aria-hidden="true"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="literal"
                >
                  /
                </div>
                <div
                  aria-describedby="react-aria-description-2 react-aria-:r4:"
                  aria-invalid="true"
                  aria-label="day, End Date, "
                  aria-labelledby="react-aria-:rv: react-aria-:r1:"
                  aria-valuemax="31"
                  aria-valuemin="1"
                  aria-valuenow="30"
                  aria-valuetext="30"
                  autocorrect="off"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  contenteditable="true"
                  data-editable="true"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="day"
                  enterkeyhint="next"
                  id="react-aria-:rv:"
                  inputmode="numeric"
                  role="spinbutton"
                  spellcheck="false"
                  style="caret-color: transparent;"
                  tabindex="0"
                >
                  30
                </div>
                <div
                  aria-hidden="true"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="literal"
                >
                  /
                </div>
                <div
                  aria-describedby="react-aria-description-2 react-aria-:r4:"
                  aria-invalid="true"
                  aria-label="year, End Date, "
                  aria-labelledby="react-aria-:r13: react-aria-:r1:"
                  aria-valuemax="9999"
                  aria-valuemin="1"
                  aria-valuenow="2024"
                  aria-valuetext="2024"
                  autocorrect="off"
                  class="group first:-ml-0.5 [&:not(:first-child)]:-ml-1 px-0.5 my-auto box-content tabular-nums text-start inline-block outline-none focus:shadow-sm rounded-md text-foreground-500 data-[editable=true]:text-foreground data-[editable=true]:data-[placeholder=true]:text-foreground-500 data-[invalid=true]:text-danger-300 data-[invalid=true]:data-[editable=true]:text-danger data-[invalid=true]:focus:bg-danger-400/50 dark:data-[invalid=true]:focus:bg-danger-400/20 data-[invalid=true]:data-[editable=true]:focus:text-danger focus:bg-default-400/50 data-[editable=true]:focus:text-default-foreground"
                  contenteditable="true"
                  data-editable="true"
                  data-invalid="true"
                  data-slot="segment"
                  data-type="year"
                  enterkeyhint="next"
                  id="react-aria-:r13:"
                  inputmode="numeric"
                  role="spinbutton"
                  spellcheck="false"
                  style="caret-color: transparent;"
                  tabindex="0"
                >
                  2024
                </div>
                <input
                  type="hidden"
                  value="2024-10-30"
                />
              </div>
              <button
                aria-describedby="react-aria-description-2 react-aria-:r4:"
                aria-expanded="false"
                aria-haspopup="dialog"
                aria-label="Calendar"
                aria-labelledby="react-aria-:r5: react-aria-:r1:"
                class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-tiny gap-2 rounded-full px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-transparent data-[hover=true]:bg-default/40 min-w-8 w-8 h-8 -mr-2 text-inherit"
                data-slot="selector-button"
                id="react-aria-:r5:"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="text-lg text-inherit pointer-events-none flex-shrink-0"
                  data-slot="selector-icon"
                  fill="none"
                  focusable="false"
                  height="1em"
                  role="presentation"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    d="M7.75 2.5a.75.75 0 0 0-1.5 0v1.58c-1.44.115-2.384.397-3.078 1.092c-.695.694-.977 1.639-1.093 3.078h19.842c-.116-1.44-.398-2.384-1.093-3.078c-.694-.695-1.639-.977-3.078-1.093V2.5a.75.75 0 0 0-1.5 0v1.513C15.585 4 14.839 4 14 4h-4c-.839 0-1.585 0-2.25.013z"
                    fill="currentColor"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M2 12c0-.839 0-1.585.013-2.25h19.974C22 10.415 22 11.161 22 12v2c0 3.771 0 5.657-1.172 6.828C19.657 22 17.771 22 14 22h-4c-3.771 0-5.657 0-6.828-1.172C2 19.657 2 17.771 2 14zm15 2a1 1 0 1 0 0-2a1 1 0 0 0 0 2m0 4a1 1 0 1 0 0-2a1 1 0 0 0 0 2m-4-5a1 1 0 1 1-2 0a1 1 0 0 1 2 0m0 4a1 1 0 1 1-2 0a1 1 0 0 1 2 0m-6-3a1 1 0 1 0 0-2a1 1 0 0 0 0 2m0 4a1 1 0 1 0 0-2a1 1 0 0 0 0 2"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="hidden group-data-[has-helper=true]:flex p-1 relative flex-col gap-1.5"
            data-slot="helper-wrapper"
          >
            <div
              class="text-tiny text-danger"
              data-slot="error-message"
              id="react-aria-:r4:"
            >
              Value must be 9/7/2024 or earlier.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
